### PR TITLE
Make average compatible across ruby versions

### LINKED ISSUE #34850: ActiveRecord Calculations average returns 0 instead of `nil` when there are no rows
### Steps to reproduce
```
Person.average(:age) 
```
on a table with no rows

### Expected behavior
returns `nil` https://apidock.com/rails/ActiveRecord/Calculations/average

### Actual behavior
returns `0.0`

### System configuration
**Rails version**: 5.1.6.1

**Ruby version**: 2.6.0

## Note

I think the reason could be this commit in bigdecimal https://github.com/ruby/bigdecimal/commit/3d1bdee3766b7eb611a2d845012f0eedaa3a790c#diff-a94632d0c7a3bcb163ce0724c2785ca1 

because ActiveRecord checks for `respond_to?(:to_d)` which returns false on a rails console in ruby 2.5.1 but true on 2.6.0

### PR DESCRIPTION
**Pull request by albertoalmagro**
### Summary

Since Ruby 2.6.0 `NilClass#to_d` is returning `BigDecimal` 0.0. This breaks `average` compatibility with prior Ruby versions. This patch makes `average` return `nil` in all Ruby versions when there are no rows.

### Other Information

Reverts #34601
Fixes #34850

r? @rafaelfranca

### CODE REVIEW

File: activerecord/lib/active_record/relation/calculations.rb [hunk 1]
```diff
@@ -401,11 +401,17 @@ def type_cast_calculated_value(value, type, operation = nil)
         case operation
         when "count"   then value.to_i
         when "sum"     then type.deserialize(value || 0)
-        when "average" then value.respond_to?(:to_d) ? value.to_d : value
+        when "average" then cast_average(value)
```
Comment (kamipo) [inline activerecord/lib/active_record/relation/calculations.rb]: Just enough `value&.respond_to?(:to_d) ? value.to_d : value` or `value && value.respond_to?(:to_d) ? value.to_d : value`.

```suggestion
        when "average" then value && value.respond_to?(:to_d) ? value.to_d : value
```

Comment (albertoalmagro) [reply activerecord/lib/active_record/relation/calculations.rb]: Thanks for the suggestion @kamipo change applied. 🙏

**Comment by kamipo** [comment]
Returning `nil` for `average` no records is preferable to me.

The reason that I pasted the all operation cases in https://github.com/rails/rails/pull/34601#discussion_r238217343 was for `when "sum"`.

https://github.com/rails/rails/blob/e7717ec7a09ca83f92a9819303f7a93e038a2679/activerecord/lib/active_record/relation/calculations.rb#L400-L405

Databases also returns `nil` for `sum` no records, but now Active Record regards no records `sum` as `0` instead of `nil`.

I believe that is historical reason, but it is a good timing to discuss the odd inconsistency for no records `sum` and `average`.

**Comment by rafaelfranca** [comment]
I think it makes sense to `sum` return `0` and average return `nil`. For sum it doesn't make difference if there is no records or if there is many that the sum returns 0, but for average it does make a difference since you don't know if 0 is because you have no record or because you have two records with `-10` and `10` as value.

**Comment by kamipo** [comment]
I don't know I get your point, you means that the difference (1) and (3) is less important than the difference (2) and (4), right?

```
root@localhost [test] > create table t (c int);
Query OK, 0 rows affected (0.03 sec)

root@localhost [test] > select sum(c) from t; -- (1)
+--------+
| sum(c) |
+--------+
|   NULL |
+--------+
1 row in set (0.00 sec)

root@localhost [test] > select avg(c) from t; -- (2)
+--------+
| avg(c) |
+--------+
|   NULL |
+--------+
1 row in set (0.00 sec)

root@localhost [test] > insert into t values (10), (-10);
Query OK, 2 rows affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

root@localhost [test] > select sum(c) from t; -- (3)
+--------+
| sum(c) |
+--------+
|      0 |
+--------+
1 row in set (0.00 sec)

root@localhost [test] > select avg(c) from t; -- (4)
+--------+
| avg(c) |
+--------+
| 0.0000 |
+--------+
1 row in set (0.00 sec)
```

**Comment by rafaelfranca** [comment]
Yes. I can't think in a case where I need to know the sum of all records is 0 because there are no records or because all records are 0.

**Pull request merged**